### ISSUE TITLE
Fix failures with logging enabled in OpenZeppelin Contracts

### ISSUE THREAD

**Issue by agostbiro** [NomicFoundation/edr#302]
**Steps to reproduce:**

1. Check out https://github.com/OpenZeppelin/openzeppelin-contracts at the latest master branch
1. [Enable logging](https://hardhat.org/hardhat-network/docs/reference#loggingenabled) in Open Zeppelin Contracts
1. Check out `d351c2b50` in Hardhat
1. Change the Hardhat dependency in Open Zeppelin Contracts to your local Hardhat version
1. Run `for i in {1..10}; do npx hardhat test --bail; done`

**Expected result**

The tests pass.

**Actual result**

Sporadic failures, e.g: https://pastebin.com/6qhHMfcv

**Definition of done**

The test command succeeds.

**Comment by Wodann** [comment]
I'm not able to reproduce this.

**Comment by agostbiro** [comment]
I'm not able to reproduce it anymore either, so I'm going to close the issue. 😅  Some learnings below from trying to identify the root cause.

Based on my understanding of the JS event loop, given code like this: `Promise.all([sendTx1(), sendTx2()])`,
 if the functions don’t do any i/o, then the execution order will be deterministic, because the tasks are added to a queue which preserves ordering and the only way for `sendTx2` to resolve before `sendTx1` is if they do network requests for example and the second finishes before the first. But if all the work is CPU bound, then I think `sendTx1` always resolves before `sendTx2` because `sendTx1` will always be in ready state in the queue before `sendTx2`, because it’s evaluated first. 

So if there is code like this with a local blockchain:
`Promise.all[(snapshot(), sendTx(), sendCallToAssertSomething(), revert()]`
Imo this will always execute in the order written in the `ethereumjs` branch.

With `edr` we’re using a FIFO lock, so in theory it’d preserve the order, but in practice it actually doesn’t, because the locking is happening in a separate thread for each request outside of the event loop and the order in which those threads execute is undefined. So we might be breaking something here.

I wrote this snippet to verify my assertion about deterministic execution of CPU tasks:
```js
async function one() {
	console.log("1");
}

async function two() {
	console.log("2");
}

async function main() {
	for (var i = 0; i < 1000000; i += 1) {
		await Promise.all([one(), two()])
	}
}

main();
```

It never prints “1\n1” or “2\n2"

**Comment by Wodann** [comment]
The reason you never seen `1\n1` or `2\n2` is that only the `await Promise.all(..)` needs to be resolved before the for-loop continues.

The only thing that could be intermingled is if you have:

```ts
async function one_two() {
    console.log("1");
    await something();
    console.log("2");
}

async function three_four() {
    console.log("3");
    await something();
    console.log("4");
}

async function main() {
	for (var i = 0; i < 1000000; i += 1) {
		await Promise.all([one_two(), three_four()])
	}
}

main();
```

You might see "3" and "4" before you see "2"

**Comment by agostbiro** [comment]
We've managed to reproduce this in the end with the following command:

```
for i in {1..5}; do rm -rf node_modules && npm i --registry=http://localhost:4873 && npx hardhat compile && time npx hardhat test --bail --grep Access || break; done
```

Nuking `node_modules` and reinstalling from scratch is important.

Turns out that the cause is errors in the OZ test suite that cause indetirminism. These were fixed by @fvictorio : https://github.com/fvictorio/openzeppelin-solidity/tree/async-fixes-in-tests

With @fvictorio 's fix there are some assertions that are failing now, but they fail both with EDR and EthereumJS so they're safe to ignore.
